File: Backjoon/Python/2644_촌수계산.py
"""
 촌수계산
 https://www.acmicpc.net/problem/2644
"""

input()
x,y = map(int,input().split())
parents = {}
for _ in range(int(input())):
  p, c = map(int,input().split())
  parents[c] = p

result = -1
xParents = [x]
yParents = [y]
xParent = parents[x] if x in parents else 0
yParent = parents[y] if y in parents else 0
while xParent != 0 or yParent != 0:
    if xParent != 0:
        xParents.append(xParent)
        xParent = parents[xParent] if xParent in parents else 0
    if yParent != 0:
        yParents.append(yParent)
        yParent = parents[yParent] if yParent in parents else 0
    i = set(xParents).intersection(set(yParents))
    if i:
        v = i.pop()
        result = xParents.index(v) + yParents.index(v)
        break
print(result)

File: Backjoon/Sliver/2644_촌수계산.swift
/*
 촌수계산
 https://www.acmicpc.net/problem/2644
*/

let n = Int(readLine()!)!
let persons = readLine()!.split(separator: " ").map { Int("\($0)")! }
var parents: [Int:Int] = [:]
for  _ in 0..<Int(readLine()!)! {
    let pc = readLine()!.split(separator: " ").map { Int("\($0)")! }
    parents[pc[1]] = pc[0]
}

var result = -1
var xParents: [Int] = [persons[0]]
var yParents: [Int] = [persons[1]]
var xParent: Int? = parents[persons[0]]
var yParent: Int? = parents[persons[1]]
while xParent != nil || yParent != nil {
    if let value = xParent {
        xParents.append(value)
        xParent = parents[value]
    }
    if let value = yParent {
        yParents.append(value)
        yParent = parents[value]
    }
    let t = Set(xParents).intersection(Set(yParents))
    if let value = t.first {
        result = xParents.firstIndex(of: value)! + yParents.firstIndex(of: value)!
        break
    }
}
print(result)
